### ISSUE TITLE
HTTP/2 fallback

### ISSUE THREAD

**Issue by mirjak** [ietf-wg-masque/draft-ietf-masque-ip-proxy-reqs#24]
The requirements draft says

`The protocol SHOULD also support HTTP/2 {{!H2=RFC7540}} as a fallback when UDP is blocked on the network path.`

and the charter says

`The working group will consider fallback to versions of HTTPS that operate over TCP as a mitigation to UDP or HTTP/3 blocking.`

However both of these sentences aren't hard requirement and I think we should have a bit more discussion about the assumptions here.

The base assumption is that UDP might be blocked between the client and proxy. That's fine and probably a valid assumption that this can happen. However, it's less clear to me what's the appropriate thing to do in this situation. There are multiple options:

- you give up...
- you directly connect to the target service without the proxy. This is not possible in case where the proxy in needed to enable connectivity but there might be other proxy services that are more optional.
- You fall back to HTTP CONNECT. This means the target server would need to support TCP; however, that's probably still a valid assumption these days.
- you fall back to Masque over H2, meaning you use one TCP tunnel to multiplex all traffic to the proxy.

So the last case is really only needed if you really, really need to connect to the target server over a proxy and the target server does not speak TCP.

My assumption is that, given we have to support stream-based transport anyway, falling back to H2 might not be too complicated, however, we shouldn't constrain ourselves in the design only in order to realise this fallback (with the base protocol).

**Comment by achernya** [comment]
I believe a MASQUE fallback to H2 is desirable. I do not believe this imposes any significant constraints on the base protocol other than a requirement to support multiple Data Transports, which are already in the requirements document. @DavidSchinazi any thoughts?

**Comment by mirjak** [comment]
Can you future say why you think it is desirable? I think that the important information for this discussion.

I don't disagree that is might not impose any significant constraints, however, that depends on the actually solution and not a good reason to have a requirement. My question basically is if we want to weaken this requirement as it seem less important compared to others, and if we maybe also want to add a bit more text about other options to fallback...?

**Comment by DavidSchinazi** [comment]
It's say it's more than desirable - it's required. Here are scenarios where you cannot connect directly to the server if the proxied connection fails:
1. if the server is only accessible via VPN
2. if the client uses IP proxying to protect its privacy from its ISP

Based on those, we need the protocol to function when UDP is blocked, and therefore operate over HTTTP/2.

**Comment by achernya** [comment]
Thanks @DavidSchinazi for outlining this. In particular, we've heard from some potential partners that their network infrastructure blocks udp/443 but tcp/443 is allowed, so having the fallback is important for functionality while they work through the process to support more modern protocols.

Given that this seems to be a pretty clear requirement for two use cases, I recommend we close out this issue, as there do not seem to be any changes to make to the current draft.

**Comment by mirjak** [comment]
@DavidSchinazi there is today usually also the option to fallback to TCP and HTTP CONNECT when UDP is blocked. I'm not saying that we shouldn't consider a fallback to H2, I just want to understand if that is a strong requirement or more a nice-to-have (which might actually be easy enough to realised that we well might end up with support it).

@achernya not sure which use case and which partner you are talking about but if it's their own infrastructure and they want to use MASQUE in an efficient way, they should probably take action to unblock upd/443.

**Comment by achernya** [comment]
@mirjak That's the point: the partners prioritize "functioning" over "efficient", and it's a lot easier to direct a partner to an FAQ on how to make something work better than to require it to work at all.

There is also prior art here; I believe Cisco AnyConnect uses TLS over TCP if DTLS is not available.

To me, this makes H2 fallback a hard requirement.

**Comment by chris-wood** [comment]
Based on the interim meeting and April 27 consensus call, there is consensus to merge #28 and close this issue. Editors, can you please do so?
